File: iris-ld/iris-ld-proofs/src/main/java/city/smartb/iris/ldproof/signer/JoseJWSSigner.java
package city.smartb.iris.ldproof.signer;

import city.smartb.iris.crypto.dsl.signer.Signer;
import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.JWSHeader;
import com.nimbusds.jose.JWSSigner;
import com.nimbusds.jose.crypto.impl.BaseJWSProvider;
import com.nimbusds.jose.util.Base64URL;

import java.security.GeneralSecurityException;
import java.util.Collections;

public class JoseJWSSigner extends BaseJWSProvider implements JWSSigner {

	private final Signer signer;

	public JoseJWSSigner(Signer signer, JWSAlgorithm algorithm) {
		super(Collections.singleton(algorithm));
		this.signer = signer;
	}

	@Override
	public Base64URL sign(final JWSHeader header, final byte[] signingInput) throws JOSEException {
		if (! this.supportedJWSAlgorithms().contains(header.getAlgorithm())) throw new JOSEException("Unexpected algorithm: " + header.getAlgorithm());
		try {
			return Base64URL.encode(this.signer.sign(signingInput));
		} catch (GeneralSecurityException ex) {
			throw new JOSEException(ex.getMessage(), ex);
		}
	}

}

File: iris-ld/iris-ld-proofs/src/main/kotlin/city/smartb/iris/ld/ldproof/signer/JoseJWSSigner.kt
package city.smartb.iris.ld.ldproof.signer

import city.smartb.iris.crypto.dsl.signer.Signer
import com.nimbusds.jose.JOSEException
import com.nimbusds.jose.JWSAlgorithm
import com.nimbusds.jose.JWSHeader
import com.nimbusds.jose.crypto.impl.BaseJWSProvider
import com.nimbusds.jose.util.Base64URL
import java.security.GeneralSecurityException

class JoseJWSSigner(private val signer: Signer, algorithm: JWSAlgorithm?) : BaseJWSProvider(setOf(algorithm)){
    @Throws(JOSEException::class)
    suspend fun sign(header: JWSHeader, signingInput: ByteArray): Base64URL {
        if (!supportedJWSAlgorithms().contains(header.algorithm)) throw JOSEException("Unexpected algorithm: " + header.algorithm)
        return try {
            Base64URL.encode(signer.sign(signingInput))
        } catch (ex: GeneralSecurityException) {
            throw JOSEException(ex.message, ex)
        }
    }
}
